### PR TITLE
Support multiple Redis clients

### LINKED ISSUE #11558: Support multiple redis clients
**Description**
Support multiple redis clients to different redis instances.

**Implementation ideas**
I assume the implementation should be similar to https://quarkus.io/guides/datasource#multiple-datasources or at least follow the same pattern.

### LINKED ISSUE #11986: Problem connecting to Redis Cluster on sentinel type
**Describe the bug**
Cannot connect Redis client on with sentinel type. This is a bug because is not supporting this type of configuration.

**Expected behavior**
Can ping /use the redis clients with sentinel type successfully.

**Actual behavior**
Because current `RedisAPIProducer.java#buildConnectionString()` is doing :
```java
      builder.append(address.getHostString());
        builder.append(COLON);
        builder.append(address.getPort());
        builder.append(SLASH);
        builder.append(config.database);
        }`
```
producing : redis://sentinel1:26379/0 -> with database config append,
on sentinel mode we cannot have the endpoint with database defined causing the following error:

`-ERR unknown command `SELECT`, with args beginning with: `0`,`

**To Reproduce**
1. Setup a redis cluster using sentinel configuration
2. change quarkus config with sentinel hosts and type: sentinel
3. do a ping with provided client

**Configuration**
```properties
#Redis
quarkus.redis.enabled=true
quarkus.redis.password=
quarkus.redis.hosts=sentinel1:26379
quarkus.redis.client-type=sentinel
```

**Additional context**
So a tip to solve this is just a simple refactor on `RedisAPIProducer.java#buildConnectionString()` :
```java
builder.append(address.getHostString());
        builder.append(COLON);
        builder.append(address.getPort());

        if (config.clientType != RedisClientType.SENTINEL) {
            builder.append(SLASH);
            builder.append(config.database);
        }
producing: `redis://sentinel1:26379` -> the expected endpoint for a sentinel host
```

### PR DESCRIPTION
**Pull request by machi1990**
This is into two separate commits as the first one only add some additional configuration options.

The second commit is the "real" subject of this PR.

Fixes https://github.com/quarkusio/quarkus/issues/11558

### CODE REVIEW

File: extensions/redis-client/deployment/src/main/java/io/quarkus/redis/client/deployment/RedisClientProcessor.java [hunk 1]
```diff
@@ -39,4 +72,101 @@ HealthBuildItem addHealthCheck(RedisBuildTimeConfig buildTimeConfig) {
...
+        for (AnnotationInstance annotation : clientAnnotations) {
+            clientNames.add(annotation.value().asString());
+        }
+
+        for (String clientName : clientNames) {
+            syntheticBeans.produce(createBlockingSyntheticBean(recorder, clientName));
+            syntheticBeans.produce(createReactiveSyntheticBean(recorder, clientName));
+            syntheticBeans.produce(createMutinyRedisApiSyntheticBean(recorder, clientName));
+            syntheticBeans.produce(createMutinyRedisClientSyntheticBean(recorder, clientName));
+            syntheticBeans.produce(createVertxRedisAPISyntheticBean(recorder, clientName));
```
Comment (gsmet) [inline extensions/redis-client/deployment/src/main/java/io/quarkus/redis/client/deployment/RedisClientProcessor.java:95]: The one two lines above is called `Api` not `API`. Could we be consistent?

Comment (machi1990) [reply extensions/redis-client/deployment/src/main/java/io/quarkus/redis/client/deployment/RedisClientProcessor.java]: Good point, thanks for spotting this. I had this mixed up all over the codebase and I've applied some clean up.

**Comment by vitor-cardoso-tdx** [comment]
Looks good , just a question can we port this improvement to 1.7.x some next version because right now I cannot use the sentinel mode: https://github.com/quarkusio/quarkus/issues/11986

**Comment by gsmet** [comment]
@machi1990 I think we need a separate PR for the Sentinel fix targeting 1.8 as it looks like something we will want in 1.8.1.

**Comment by machi1990** [comment]
> @machi1990 I think we need a separate PR for the Sentinel fix targeting 1.8 as it looks like something we will want in 1.8.1.

To fix the issue, I had to refactor the redis configuration a little. Is that something that's desirable for 1.8.1?

**Comment by machi1990** [comment]
> @machi1990 yeah, probably not a good idea.

Great. I'll be inclined for the fix to go in with this PR at a later release.

File: docs/src/main/asciidoc/redis.adoc [hunk 2]
```diff
@@ -544,6 +544,40 @@ So when you access the `/health/ready` endpoint of your application you will hav
 
 This behavior can be disabled by setting the `quarkus.redis.health.enabled` property to `false` in your `application.properties`.
 
+== Multiple Redis Clients
+
+Redis allows you to configure multiple clients.
```
Comment (gsmet) [inline docs/src/main/asciidoc/redis.adoc:549]:
```suggestion
The Redis extension allows you to configure multiple clients.
```
maybe?

File: docs/src/main/asciidoc/redis.adoc [hunk 3]
```diff
@@ -544,6 +544,40 @@ So when you access the `/health/ready` endpoint of your application you will hav
...
+quarkus.redis.second.hosts = localhost:6379
+----
+
+Notice there's an extra bit in the key (the `second` segment).
+The syntax is as follows: `quarkus.redis.[optional name.][redis configuration property]`.
+If the name is omitted, it configures the default client.
+
+=== Named Redis client Injection
+
+When using multiple clients, each `RedisClient`, you can select the client to inject using the `io.quarkus.redis.client.RedisClientName` qualifier.
```
Comment (gsmet) [inline docs/src/main/asciidoc/redis.adoc:564]: There's something weird with this sentence, isn't it? The `each RedisClient` part sounds weird.

File: extensions/redis-client/runtime/src/main/java/io/quarkus/redis/client/RedisClientName.java [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+
+import java.lang.annotation.Documented;
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Retention;
+import java.lang.annotation.Target;
+
+import javax.inject.Qualifier;
+
+/**
+ * Marker annotation to select redis client
```
Comment (gsmet) [inline extensions/redis-client/runtime/src/main/java/io/quarkus/redis/client/RedisClientName.java:13]:
```suggestion
 * Marker annotation to select the Redis client.
```

File: extensions/redis-client/runtime/src/main/java/io/quarkus/redis/client/RedisClientName.java [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+ * Use name parameter to select it
+ *
+ * For example, if a redis connection is configured like so in {@code application.properties}:
```
Comment (gsmet) [inline extensions/redis-client/runtime/src/main/java/io/quarkus/redis/client/RedisClientName.java:16]:
```suggestion
 * For example, if a Redis connection is configured like so in {@code application.properties}:
```

File: extensions/redis-client/runtime/src/main/java/io/quarkus/redis/client/RedisClientName.java [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+import java.lang.annotation.Documented;
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Retention;
+import java.lang.annotation.Target;
+
+import javax.inject.Qualifier;
+
+/**
+ * Marker annotation to select redis client
+ * Use name parameter to select it
```
Comment (gsmet) [inline extensions/redis-client/runtime/src/main/java/io/quarkus/redis/client/RedisClientName.java:14]: Let's just drop this sentence, it's just noise.

File: extensions/redis-client/runtime/src/main/java/io/quarkus/redis/client/runtime/RedisClientUtil.java [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+import java.net.URI;
+import java.util.Set;
+
+import io.quarkus.redis.client.runtime.RedisConfig.RedisConfiguration;
+import io.quarkus.runtime.configuration.ConfigurationException;
+import io.vertx.redis.client.RedisClientType;
+import io.vertx.redis.client.RedisOptions;
+
+public class RedisClientUtil {
+    public static final String DEFAULT_CLIENT = "__redis_default_client__";
```
Comment (gsmet) [inline extensions/redis-client/runtime/src/main/java/io/quarkus/redis/client/runtime/RedisClientUtil.java:12]: In the other extensions, we call it `<default>`.

File: extensions/redis-client/runtime/src/main/java/io/quarkus/redis/client/runtime/RedisConfig.java [hunk 8]
```diff
@@ -1,77 +1,136 @@
...
+     * The default redis client
      */
-    @ConfigItem
-    public Optional<String> password;
+    @ConfigItem(name = ConfigItem.PARENT)
+    public RedisConfiguration defaultClient;
 
     /**
-     * The redis hosts
+     * Configures additional redis client connections.
```
Comment (gsmet) [inline extensions/redis-client/runtime/src/main/java/io/quarkus/redis/client/runtime/RedisConfig.java:27]:
```suggestion
     * Configures additional Redis client connections.
```

File: extensions/redis-client/runtime/src/main/java/io/quarkus/redis/client/runtime/RedisConfig.java [hunk 9]
```diff
@@ -1,77 +1,136 @@
...
+     * example:
+     * <p>
+     *
+     * <pre>
+     * quarkus.redis.client1.hosts = redis://localhost:6379
+     * quarkus.redis.client2.hosts = redis://localhost:6380
+     * </pre>
+     * <p>
+     * And then use annotations above the instances of RedisClient to indicate which instance we are going to use
+     * <p>
```
Comment (gsmet) [inline extensions/redis-client/runtime/src/main/java/io/quarkus/redis/client/runtime/RedisConfig.java:39]: Was it copy/pasted from elsewhere? All this Javadoc block could use some rework.

**Review by gsmet — CHANGES_REQUESTED** [review]
Looks mostly good. I added some comments.

File: docs/src/main/asciidoc/redis.adoc (hunk 3 above)
Comment (machi1990) [reply docs/src/main/asciidoc/redis.adoc:564]: Yeap. That part should not be there.

File: extensions/redis-client/runtime/src/main/java/io/quarkus/redis/client/runtime/RedisClientUtil.java (hunk 7 above)
Comment (machi1990) [reply extensions/redis-client/runtime/src/main/java/io/quarkus/redis/client/runtime/RedisClientUtil.java:12]: Sure, let's be consistent here too. I'll change it.

File: extensions/redis-client/runtime/src/main/java/io/quarkus/redis/client/runtime/RedisConfig.java (hunk 9 above)
Comment (machi1990) [reply extensions/redis-client/runtime/src/main/java/io/quarkus/redis/client/runtime/RedisConfig.java:39]: I'll push some updates. Let me know what you think.

**Comment by machi1990** [comment]
@gsmet thanks for the review. I've addressed your comments. Let me know what you think.

File: extensions/redis-client/runtime/src/main/java/io/quarkus/redis/client/RedisClientName.java [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+
+import java.lang.annotation.Documented;
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Retention;
+import java.lang.annotation.Target;
+
+import javax.inject.Qualifier;
+
+/**
+ * Marker annotation to select the redis client
```
Comment (gsmet) [inline extensions/redis-client/runtime/src/main/java/io/quarkus/redis/client/RedisClientName.java:13]:
```suggestion
 * Marker annotation to select the Redis client.
```
(it's minor but this is public API so let's get it right)

File: extensions/redis-client/runtime/src/main/java/io/quarkus/redis/client/RedisClientName.java [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+ */
+@Target({ ElementType.TYPE, ElementType.METHOD, ElementType.FIELD })
+@Retention(RUNTIME)
+@Documented
+@Qualifier
+public @interface RedisClientName {
+    /**
+     * Specify the cluster name of the connection.
+     *
+     * @return the value
```
Comment (gsmet) [inline extensions/redis-client/runtime/src/main/java/io/quarkus/redis/client/RedisClientName.java:37]: Let's drop this.

File: extensions/redis-client/runtime/src/main/java/io/quarkus/redis/client/RedisClientName.java [hunk 12]
```diff
@@ -0,0 +1,40 @@
...
+ *     RedisClient client;
+ * </pre>
+ */
+@Target({ ElementType.TYPE, ElementType.METHOD, ElementType.FIELD })
+@Retention(RUNTIME)
+@Documented
+@Qualifier
+public @interface RedisClientName {
+    /**
+     * Specify the cluster name of the connection.
```
Comment (gsmet) [inline extensions/redis-client/runtime/src/main/java/io/quarkus/redis/client/RedisClientName.java:35]:
```suggestion
     * The Redis client name.
```

File: extensions/redis-client/runtime/src/main/java/io/quarkus/redis/client/runtime/RedisConfig.java [hunk 13]
```diff
@@ -1,77 +1,140 @@
...
-    @ConfigItem
-    public Optional<String> password;
+    @ConfigItem(name = ConfigItem.PARENT)
+    public RedisConfiguration defaultClient;
 
     /**
-     * The redis hosts
+     * Configures additional Redis client connections.
+     * <p>
+     * Each client have a unique identifier which must be identified to select the right connection.
```
Comment (gsmet) [inline extensions/redis-client/runtime/src/main/java/io/quarkus/redis/client/runtime/RedisConfig.java:31]:
```suggestion
     * Each client has a unique identifier which must be identified to select the right connection.
```

File: extensions/redis-client/runtime/src/main/java/io/quarkus/redis/client/runtime/RedisConfig.java [hunk 14]
```diff
@@ -1,77 +1,140 @@
...
+     * <p>
+     *
+     * <pre>
+     * quarkus.redis.client1.hosts = redis://localhost:6379
+     * quarkus.redis.client2.hosts = redis://localhost:6380
+     * </pre>
+     * <p>
+     * And then use the {@link RedisClientName} to select the {@link RedisClient} or
+     * {@link io.quarkus.redis.client.reactive.ReactiveRedisClient}
+     * we are going to use.
```
Comment (gsmet) [inline extensions/redis-client/runtime/src/main/java/io/quarkus/redis/client/runtime/RedisConfig.java:42]:
```suggestion
     * And then use the {@link RedisClientName} annotation to select the {@link RedisClient} or
     * {@link io.quarkus.redis.client.reactive.ReactiveRedisClient}.
```

File: extensions/redis-client/runtime/src/main/java/io/quarkus/redis/client/runtime/RedisAPIProducer.java [hunk 15]
```diff
@@ -1,137 +1,55 @@
...
 import io.quarkus.redis.client.RedisClient;
 import io.quarkus.redis.client.reactive.ReactiveRedisClient;
-import io.quarkus.runtime.configuration.ConfigurationException;
 import io.vertx.core.Vertx;
 import io.vertx.redis.client.Redis;
 import io.vertx.redis.client.RedisAPI;
-import io.vertx.redis.client.RedisClientType;
 import io.vertx.redis.client.RedisOptions;
 
 class RedisAPIProducer {
```
Comment (gsmet) [inline extensions/redis-client/runtime/src/main/java/io/quarkus/redis/client/runtime/RedisAPIProducer.java:16]: Is there a reason why we are not being consistent and use the synthetic bean pattern for this too?

**Review by gsmet — CHANGES_REQUESTED** [review]
Added some additional suggestions and a question.

File: extensions/redis-client/runtime/src/main/java/io/quarkus/redis/client/runtime/RedisAPIProducer.java (hunk 15 above)
Comment (machi1990) [reply extensions/redis-client/runtime/src/main/java/io/quarkus/redis/client/runtime/RedisAPIProducer.java:16]: You mean drop this build item  https://github.com/quarkusio/quarkus/blob/6a7ba55365abf623a196fe106800c490c88fe6ac/extensions/redis-client/deployment/src/main/java/io/quarkus/redis/client/deployment/RedisClientProcessor.java#L59-L64  and use a synthetic bean instead? 

If so, I saw no need of the synthetic bean for this one.

**Comment by machi1990** [comment]
@gsmet thanks for the suggested fixes which I've applied. I've also answered https://github.com/quarkusio/quarkus/pull/11908#discussion_r490144731, let me know what you think.

File: extensions/redis-client/runtime/src/main/java/io/quarkus/redis/client/runtime/RedisAPIProducer.java (hunk 15 above)
Comment (gsmet) [reply extensions/redis-client/runtime/src/main/java/io/quarkus/redis/client/runtime/RedisAPIProducer.java:16]: OK, we can live with it I suppose.

**Review by gsmet — APPROVED** [review]

**Comment by machi1990** [comment]
Merged. @gsmet thanks for the review.

**Comment by gsmet** [comment]
Thanks for your work, it's a nice addition!

**Comment by geoand** [comment]
This is a breaking change, right?

**Comment by machi1990** [comment]
> This is a breaking change, right?

Right. See https://github.com/quarkusio/quarkus/pull/11908#issuecomment-689527908.

**Comment by gsmet** [comment]
Well, there's a difference between refactoring the config and breaking the config :).

What's the rationale for removing the `ssl` bit and include everything in the `host`? Also, could we be a bit better on that if we really need it i.e. support the old config for a while but raise a warning?

**Comment by machi1990** [comment]
> Well, there's a difference between refactoring the config and breaking the config :).

Right. 

> What's the rationale for removing the `ssl` bit and include everything in the `host`? Also, could we be a bit better on that if we really need it i.e. support the old config for a while but raise a warning?

Breaking up the connection string into several parts worked nice with a standalone redis client type but not so well with cluster or sentinel.  There was no point of exposing the property and doing construction of the connection string in the extension while the user can supply the connection string.  As for `ssl`, the scheme would have to `rediss`.   

> Also, could we be a bit better on that if we really need it i.e. support the old config for a while but raise a warning?

I agree and I would have done it this way really.

**Comment by machi1990** [comment]
> This is a breaking change, right?

Quick update regarding this: In the coming days, I'll open a followup PR to provide backward compatibility for the removed configuration properties for at least one release. This will give users ample time to move to the new way of configuring things.

**Comment by geoand** [comment]
Thanks!

**Pull request merged**